hibernate-core/src/main/java/org/hibernate/engine/internal/TwoPhaseLoad.java

```diff
@@ -150,6 +151,11 @@ private static void doInitializeEntity(
 			if ( value!=LazyPropertyInitializer.UNFETCHED_PROPERTY && value!= PropertyAccessStrategyBackRefImpl.UNKNOWN ) {
 				hydratedState[i] = types[i].resolve( value, session, entity );
 			}
+			else if ( types[i].isCollectionType() ) {
```

Comment:
IIUC, this is specifically the case where bytecode-enhanced entity has a lazy collection. The fact that the resolved, uninitialized PersistentCollection should not be assigned back to hydratedState is an implementation detail.

I tried adding the assignment to hydratedState and stepping through the test case in this PR to confirm that the collection gets initialized. I did not find that the collection gets initialized; instead I see an inconsistency that `toRemoveGarage.cars` is set to an uninitialized `PersistantCollection`, but `toRemoveGarage.$$_hibernate_attributeInterceptor.initializedLazyFields contains "cars".

With this inconsistency, Hibernate.isPropertyInitialized( toRemoveGarage, "cars" ) will return true (even though the initialized property is actually an uninitialized collection).

It would be nice to not have such implementation details leaking into this code. If that can't be avoided, then there should be comments that call out the specific case and the necessary implementation detail.